taichi/backends/wasm/codegen_wasm.cpp

```diff
@@ -84,28 +85,181 @@ class CodeGenLLVMWASM : public CodeGenLLVM {
+                                  llvm::Function::ExternalLinkage,
+                                  task_kernel_name, module.get());
+
+    for (auto &arg : func->args()) {
+      kernel_args.push_back(&arg);
+    }
+    kernel_args[0]->setName("context");
+
+    if (kernel_argument_by_val())
+      func->addParamAttr(0, llvm::Attribute::ByVal);
+
+    // entry_block has all the allocas
+    this->entry_block = llvm::BasicBlock::Create(*llvm_context, "entry", func);
+
+    // The real function body
+    func_body_bb = llvm::BasicBlock::Create(*llvm_context, "body", func);
+    builder->SetInsertPoint(func_body_bb);
+    return task_kernel_name;
+  }
+
+  void finalize_taichi_kernel_function() {
+    builder->CreateRetVoid();
+
+    // entry_block should jump to the body after all allocas are inserted
+    builder->SetInsertPoint(entry_block);
+    builder->CreateBr(func_body_bb);
+
+    if (prog->config.print_kernel_llvm_ir) {
+      static FileSequenceWriter writer("taichi_kernel_generic_llvm_ir_{:04d}.ll",
+                                       "unoptimized LLVM IR (generic)");
+      writer.write(module.get());
+    }
+    TI_ASSERT(!llvm::verifyFunction(*func, &llvm::errs()));
+    // TI_INFO("Kernel function verified.");
+  }
+
+  std::string create_taichi_get_root_address_function() {
+    auto task_function_type =
+        llvm::FunctionType::get(llvm::Type::getInt32Ty(*llvm_context),
+                                {llvm::PointerType::get(context_ty, 0)}, false);
+    auto task_kernel_name = fmt::format("get_root_address");
+    auto func = llvm::Function::Create(task_function_type,
+                                       llvm::Function::ExternalLinkage,
+                                       task_kernel_name, module.get());
+
+    std::vector<llvm::Value *> kernel_args;
+    for (auto &arg : func->args()) {
+      kernel_args.push_back(&arg);
+    }
+    kernel_args[0]->setName("context");
+
+    auto entry_block = llvm::BasicBlock::Create(*llvm_context, "entry", func);
+    auto func_body_bb = llvm::BasicBlock::Create(*llvm_context, "body", func);
+    builder->SetInsertPoint(func_body_bb);
+
+    // memory reserved for Context object shouldn't be polluted
+    llvm::Value *runtime_ptr = create_call("Context_get_runtime", {kernel_args[0]});
+    llvm::Value *runtime = builder->CreateBitCast(
+        runtime_ptr, llvm::PointerType::get(get_runtime_type("LLVMRuntime"), 0));
+    llvm::Value *root_ptr = create_call("LLVMRuntime_get_ptr_root", {runtime});
+    llvm::Value *root_address = builder->CreatePtrToInt(
+        root_ptr, llvm::Type::getInt32Ty(*llvm_context));
+    builder->CreateRet(root_address);
+
+    builder->SetInsertPoint(entry_block);
+    builder->CreateBr(func_body_bb);
+
+    TI_ASSERT(!llvm::verifyFunction(*func, &llvm::errs()));
+    // TI_INFO("Kernel function verified.");
+    return task_kernel_name;
+  }
+
+  std::string create_taichi_set_root_function() {
```

Comment:
nit: Document how this function works and why we need it? Also we may also need to set `context_ptr` to `0`